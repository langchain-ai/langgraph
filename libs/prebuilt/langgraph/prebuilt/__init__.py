"""langgraph.prebuilt exposes a higher-level API for creating and executing agents and tools."""

from langgraph.prebuilt.chat_agent_executor import create_react_agent
from langgraph.prebuilt.tool_node import (
    InjectedState,
    InjectedStore,
    ToolNode,
    tools_condition,
)
from langgraph.prebuilt.tool_validator import ValidationNode

__all__ = [
    "create_react_agent",
    "ToolNode",
    "tools_condition",
    "ValidationNode",
    "InjectedState",
    "InjectedStore",
]
