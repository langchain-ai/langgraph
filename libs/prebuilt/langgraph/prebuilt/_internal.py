from typing import Any, Literal, TypedDict

from langchain_core.messages import ToolCall


class ToolCallWithContext(TypedDict):
    """ToolCall with additional context for graph state.

    This is an internal data-structure meant to help the ToolNode accept
    tools calls with additional context (e.g. state) when dispatched using the
    `Send` API.

    The Send API is used in create_react_agent to be able to distribute the tool
    calls in parallel and support human-in-the-loop workflows where graph execution
    may be paused for an indefinite time.
    """

    tool_call: ToolCall
    __type: Literal["tool_call_with_context"]
    """Type to parameterize the payload.
    
    Using "__" as a prefix to be defensive against potential name collisions with
    regular user state.
    """
    state: Any
    """The state is provided as additional context."""
