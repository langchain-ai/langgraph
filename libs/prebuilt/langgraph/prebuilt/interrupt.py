from copy import deepcopy
from typing import Any, Literal, Optional, Union, cast

from langchain_core.messages import ToolCall, ToolMessage
from typing_extensions import TypedDict

from langgraph.types import Command, interrupt
from langgraph.utils.runnable import RunnableCallable


class HumanInterruptConfig(TypedDict):
    """Configuration that defines what actions are allowed for a human interrupt.

    This controls the available interaction options when the graph is paused for human input.

    Attributes:
        allow_ignore: Whether the human can choose to ignore/skip the current step
        allow_respond: Whether the human can provide a text response/feedback
        allow_edit: Whether the human can edit the provided content/state
        allow_accept: Whether the human can accept/approve the current state
    """

    allow_ignore: bool
    allow_respond: bool
    allow_edit: bool
    allow_accept: bool


class ActionRequest(TypedDict):
    """Represents a request for human action within the graph execution.

    Contains the action type and any associated arguments needed for the action.

    Attributes:
        action: The type or name of action being requested (e.g., "Approve XYZ action")
        args: Key-value pairs of arguments needed for the action
    """

    action: str
    args: dict


class HumanInterrupt(TypedDict):
    """Represents an interrupt triggered by the graph that requires human intervention.

    This is passed to the `interrupt` function when execution is paused for human input.

    Attributes:
        action_request: The specific action being requested from the human
        config: Configuration defining what actions are allowed
        description: Optional detailed description of what input is needed

    Example:
        ```python
        # Extract a tool call from the state and create an interrupt request
        request = HumanInterrupt(
            action_request=ActionRequest(
                action="run_command",  # The action being requested
                args={"command": "ls", "args": ["-l"]}  # Arguments for the action
            ),
            config=HumanInterruptConfig(
                allow_ignore=True,    # Allow skipping this step
                allow_respond=True,   # Allow text feedback
                allow_edit=False,     # Don't allow editing
                allow_accept=True     # Allow direct acceptance
            ),
            description="Please review the command before execution"
        )
        # Send the interrupt request and get the response
        response = interrupt([request])[0]
        ```
    """

    action_request: ActionRequest
    config: HumanInterruptConfig
    description: Optional[str]


class HumanResponse(TypedDict):
    """The response provided by a human to an interrupt, which is returned when graph execution resumes.

    Attributes:
        type: The type of response:
            - "accept": Approves the current state without changes
            - "ignore": Skips/ignores the current step
            - "response": Provides text feedback or instructions
            - "edit": Modifies the current state/content
        arg: The response payload:
            - None: For ignore/accept actions
            - str: For text responses
            - ActionRequest: For edit actions with updated content
    """

    type: Literal["accept", "ignore", "response", "edit"]
    args: Union[None, str, ActionRequest]


class InterruptToolNode(RunnableCallable):
    """Prebuilt post model hook node used to enable common patterns for tool interrupts.

    For any tools with specified policies, an interrupt will be raised when the LLM returns
    a tool call for said tool. The interrupt policy will be used to determine what sort of resume logic is allowed.
    Any of the following resume patterns are supported:

    * accept: the tool call is executed as planned
    * edit: the args for the tool call are edited and then the tool call is executed
    * response: text response/feedback is fed back into the LLM
    * ignore: the current tool call is ignored / skipped

    Args:
        **interrupt_policy: a mapping of tool names to [`HumanInterruptConfig`][prebuilt.interrupt.HumanInterruptConfig] dictionaries
            specifying which interrupt patterns to enable for said tool.

        Example:
        ```python
        from langgraph.prebuilt import create_react_agent
        from langgraph.checkpoint.memory import InMemorySaver
        from langgraph.prebuilt.interrupt import HumanInterruptConfig, InterruptToolNode
        from langgraph.types import Command


        def book_hotel(hotel_name: str) -> str:
            '''Book a room at the provided hotel.'''
            # Some hotel API calls, a sensitive / expensive operation
            return f"Booked a hotel at {hotel_name}."


        agent = create_react_agent(
            "openai:gpt-4.1",
            tools=[book_hotel],
            prompt="You are a hotel booking assistant.",
            post_model_hook=InterruptToolNode(
                book_hotel=HumanInterruptConfig(
                    allow_accept=True,
                    allow_edit=True,
                    allow_ignore=True,
                    allow_respond=True,
                )
            ),
            checkpointer=InMemorySaver(),
        )

        config = {"configurable": {"thread_id": 1}}

        response = agent.invoke(
            {"messages": [{"role": "user", "content": "please book a hotel at the hilton inn in boston."}]},
            config=config,
        )

        response = agent.invoke(Command(resume={"type": "accept"}), config=config)
        ```
    """

    def __init__(self, **interrupt_policy: HumanInterruptConfig):
        super().__init__(self._func, self._afunc)
        self.interrupt_policy = interrupt_policy

    def _interrupt(
        self,
        tool_call: ToolCall,
        interrupt_config: HumanInterruptConfig,
    ) -> Union[ToolCall, ToolMessage]:
        """Interrupt before a tool call and ask for human input."""
        call_id = tool_call["id"]
        tool_name = tool_call["name"]

        request = HumanInterrupt(
            action_request=ActionRequest(
                action=tool_name,
                args=tool_call["args"],
            ),
            config=interrupt_config,
            description=f"Please review tool call for `{tool_name}` before execution.",
        )
        response = interrupt([request])

        # resume provided by agent inbox as a list
        response = response[0] if isinstance(response, list) else response

        try:
            response_type = response.get("type")
        except AttributeError:
            raise TypeError(
                f"Unexpected resume value: {response}."
                f"Expected a dict with `'type'` key."
            )

        if response_type == "accept" and interrupt_config["allow_accept"]:
            return tool_call
        elif response_type == "edit" and interrupt_config["allow_edit"]:
            return ToolCall(
                args=cast(ActionRequest, response)["args"]["args"],
                name=tool_name,
                id=call_id,
                type="tool_call",
            )
        elif response_type == "response" and interrupt_config["allow_respond"]:
            return ToolMessage(
                content=cast(str, response["args"]),
                name=tool_name,
                tool_call_id=call_id,
                status="error",
            )
        elif response_type == "ignore" and interrupt_config["allow_ignore"]:
            return ToolMessage(
                content=f"User ignored the tool call for `{tool_name}` with id {call_id}",
                name=tool_name,
                tool_call_id=call_id,
                status="success",
            )

        allowed_types = [
            type_name
            for type_name, is_allowed in {
                "accept": interrupt_config["allow_accept"],
                "edit": interrupt_config["allow_edit"],
                "response": interrupt_config["allow_respond"],
                "ignore": interrupt_config["allow_ignore"],
            }.items()
            if is_allowed
        ]

        raise ValueError(
            f"Unexpected human response: {response}. "
            f"Expected one with `'type'` in {allowed_types} based on {tool_name}'s interrupt configuration."
        )

    def _func(self, input: dict[str, Any]) -> Command:
        ai_msg = input["messages"][-1]
        tool_calls: list[ToolCall] = deepcopy(ai_msg.tool_calls) or []
        tool_messages: list[ToolMessage] = []

        for idx, tool_call in enumerate(tool_calls):
            if interrupt_config := self.interrupt_policy.get(tool_call["name"]):
                interrupt_result = self._interrupt(
                    tool_call=tool_call, interrupt_config=interrupt_config
                )

                if isinstance(interrupt_result, ToolMessage):
                    tool_messages.append(interrupt_result)
                else:
                    tool_calls[idx] = interrupt_result

        updated_ai_msg = ai_msg.copy(update={"tool_calls": tool_calls})

        # conditional routing logic for post_model_hook will direct to the tools node
        # or agent node depending on if there are pending tool calls
        return {"messages": [updated_ai_msg, *tool_messages]}

    async def _afunc(self, input: dict[str, Any]) -> Command:
        return self._func(input)
