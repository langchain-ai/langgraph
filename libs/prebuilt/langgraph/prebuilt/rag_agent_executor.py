from typing import Any, Callable, Optional, Sequence, Union, Annotated

from langchain_core.documents import Document
from langchain_core.language_models import LanguageModelLike
from langchain_core.messages import BaseMessage, AIMessage
from langchain_core.runnables import Runnable
from langchain_core.tools import BaseTool
from typing_extensions import TypedDict

from langgraph.graph import END, StateGraph, CompiledGraph
from langgraph.graph.message import add_messages
from langgraph.managed import IsLastStep, RemainingSteps
from langgraph.store.base import Checkpointer

# Placeholder for VectorStoreRetriever if it's a custom or specific class
# from langgraph.prebuilt import VectorStoreRetriever # Assuming this path, adjust if different
# For now, assume vector_store.as_retriever() or retriever_tool is used, making this import flexible

class RagState(TypedDict):
    """The state of the RAG agent."""

    messages: Annotated[Sequence[BaseMessage], add_messages]
    question: str
    original_question: str
    documents: Optional[Sequence[Document]]
    document_assessment: Optional[str]
    generation: Optional[str]
    iterations: int
    current_phase_iterations: int
    attempted_external_search: bool
    is_last_step: IsLastStep
    remaining_steps: RemainingSteps


def create_rag_agent(
    llm: LanguageModelLike,
    embedding_model: Any,
    vector_store: Any,
    retriever_tool: Optional[BaseTool] = None,
    external_search_tools: Optional[Sequence[BaseTool]] = None,
    do_document_grading: bool = True,
    do_external_search: bool = True,
    system_message_prompt: Optional[str] = (
        "You are a helpful RAG assistant. "
        "Use the retrieved documents to answer the question. "
        "If the documents are not relevant or insufficient, you can try to rephrase the question or search externally."
    ),
    checkpointer: Optional[Checkpointer] = None,
    debug: bool = False,
) -> CompiledGraph:
    """Creates a RAG agent graph.

    Args:
        llm: The language model to use.
        embedding_model: The embedding model to use (often implicitly used by vector_store or retriever_tool).
        vector_store: The vector store for primary document retrieval.
        retriever_tool: An optional pre-configured tool for document retrieval. If None, a retriever will be 
                      created from vector_store.
        external_search_tools: Optional list of tools for external search (e.g., web, ArXiv).
        do_document_grading: If True, a step will be added to grade document relevance.
        do_external_search: If True, external_search_tools will be used if initial retrieval is insufficient.
        system_message_prompt: The system prompt to use for the LLM.
        checkpointer: An optional checkpointer for persisting state.
        debug: If True, enables debug logging for the graph.

    Returns:
        A compiled LangGraph runnable for the RAG agent.
    """
    MAX_QUERY_TRANSFORMATIONS = 2  # Max attempts to rephrase a query

    _retriever: Union[BaseTool, Runnable]
    if retriever_tool is not None:
        _retriever = retriever_tool
    elif hasattr(vector_store, 'as_retriever') and callable(vector_store.as_retriever):
        # Standard LangChain vector stores have this method
        _retriever = vector_store.as_retriever()
    else:
        # Fallback for simpler vector stores or direct use, assuming they are runnable/callable
        # This part might need adjustment based on how `vector_store` is expected to be used if not a LangChain VS
        # For now, assuming it's a LangChain VectorStore or compatible retriever
        raise ValueError("vector_store must have an 'as_retriever' method if retriever_tool is not provided.")

    # 1. Define Node Functions
    def retrieve_documents_node(state: RagState) -> dict:
        print("---RETRIEVING DOCUMENTS---")
        question = state["question"]
        if state.get("original_question") is None:
            original_question = question
        else:
            original_question = state["original_question"]

        try:
            retrieved_docs = _retriever.invoke(question) # type: ignore
        except Exception as e:
            print(f"Error invoking retriever: {e}")
            retrieved_docs = []

        return {
            "documents": retrieved_docs,
            "original_question": original_question,
            "iterations": state.get("iterations", 0) + 1,
            "current_phase_iterations": 0,
            "attempted_external_search": False,
            "document_assessment": None,
        }

    def grade_retrieved_documents_node(state: RagState) -> dict:
        print("---GRADING DOCUMENTS---")
        question = state["question"]
        documents = state["documents"]

        if not documents:
            print("No documents to grade.")
            return {"document_assessment": "no_documents"}

        formatted_docs = "\n\n".join([doc.page_content for doc in documents])
        grading_prompt_template = (
            "Given the following question and retrieved documents, assess if the documents are relevant to answer the question. "
            "Respond with only 'relevant' or 'not_relevant'.\n\n"
            "Question: {question}\n\n"
            "Documents:\n{documents_text}\n\n"
            "Assessment:"
        )
        grading_prompt = grading_prompt_template.format(
            question=question, documents_text=formatted_docs
        )

        try:
            response = llm.invoke(grading_prompt) # type: ignore
            if hasattr(response, 'content'):
                assessment = response.content.strip().lower()
            elif isinstance(response, str):
                assessment = response.strip().lower()
            else:
                print(f"Unexpected LLM response type for grading: {type(response)}")
                assessment = "not_relevant"

            if assessment not in ["relevant", "not_relevant"]:
                print(f"LLM grader returned non-standard assessment: '{assessment}'. Defaulting to not_relevant.")
                assessment = "not_relevant"

        except Exception as e:
            print(f"Error invoking LLM for grading: {e}")
            assessment = "not_relevant"

        print(f"Document assessment: {assessment}")
        return {"document_assessment": assessment}

    def generate_answer_node(state: RagState) -> dict:
        print("---GENERATING ANSWER---")
        question = state["question"]
        documents = state["documents"]
        document_assessment = state.get("document_assessment")

        documents_text = ""
        use_documents = False
        if documents:
            if not do_document_grading or document_assessment == "relevant":
                documents_text = "\n\n".join([doc.page_content for doc in documents])
                use_documents = True
            elif document_assessment == "no_documents":
                documents_text = "No documents were found during retrieval."
            else:
                documents_text = "The retrieved documents were not considered relevant to the question."
        else:
            documents_text = "No documents were retrieved."

        prompt_parts = [system_message_prompt, f"User Question: {question}"]
        if use_documents:
            prompt_parts.append(f"Retrieved Documents:\n{documents_text}")
            prompt_parts.append("Based on the question and the retrieved documents, provide a comprehensive answer.")
        else:
            prompt_parts.append(f"Context: {documents_text}")
            prompt_parts.append("Please answer the question based on your general knowledge or indicate if you cannot answer without relevant documents.")
        generation_prompt = "\n\n".join(prompt_parts)

        try:
            response = llm.invoke(generation_prompt) # type: ignore
            if hasattr(response, 'content'):
                generated_answer = response.content
            elif isinstance(response, str):
                generated_answer = response
            else:
                print(f"Unexpected LLM response type for generation: {type(response)}")
                generated_answer = "Sorry, I encountered an error while generating the answer."
        except Exception as e:
            print(f"Error invoking LLM for generation: {e}")
            generated_answer = "Sorry, I encountered an error and cannot provide an answer at this time."
        print(f"Generated answer: {generated_answer[:100]}...")
        new_messages = state.get("messages", []) + [AIMessage(content=generated_answer)]

        return {
            "generation": generated_answer,
            "messages": new_messages
        }

    def transform_query_node(state: RagState) -> dict:
        print("---TRANSFORMING QUERY---")
        original_question = state["original_question"]
        current_question = state["question"]
        documents = state["documents"]
        document_assessment = state.get("document_assessment")

        transformation_context = f"The original question was: '{original_question}'."
        transformation_context += f" The last attempted query was: '{current_question}'."

        if not documents or document_assessment == "no_documents":
            transformation_context += " No documents were found with the last query."
        elif document_assessment == "not_relevant":
            transformation_context += " The retrieved documents were not relevant to the question."

        transformation_prompt_template = (
            "You are an expert at rephrasing questions to improve information retrieval.\n"
            "{context}\n\n"
            "Based on the original question and the issues encountered, provide a new, rephrased question that might yield better search results. "
            "Output ONLY the new question.\n\n"
            "New Question:"
        )
        transformation_prompt = transformation_prompt_template.format(context=transformation_context)

        try:
            response = llm.invoke(transformation_prompt) # type: ignore
            if hasattr(response, 'content'):
                new_question = response.content.strip()
            elif isinstance(response, str):
                new_question = response.strip()
            else:
                print(f"Unexpected LLM response type for query transformation: {type(response)}")
                new_question = original_question
            
            if not new_question or new_question.lower() == "new question:":
                 print("Query transformation resulted in empty or boilerplate response. Using original question.")
                 new_question = original_question

        except Exception as e:
            print(f"Error invoking LLM for query transformation: {e}")
            new_question = original_question

        print(f"Transformed query: {new_question}")
        return {
            "question": new_question,
            "documents": None,
            "document_assessment": None,
            "generation": None,
            "current_phase_iterations": state.get("current_phase_iterations", 0) + 1,
        }

    def perform_external_search_node(state: RagState) -> dict:
        print("---PERFORMING EXTERNAL SEARCH---")
        question = state["question"]
        all_external_docs: list[Document] = []

        if not external_search_tools:
            return {"documents": [], "attempted_external_search": True}

        for tool in external_search_tools:
            try:
                tool_name = tool.name if hasattr(tool, 'name') else 'UnnamedTool'
                print(f"Invoking external search tool: {tool_name}")
                tool_output = tool.invoke(question)
                
                if isinstance(tool_output, str):
                    all_external_docs.append(Document(page_content=tool_output, metadata={"source": tool_name}))
                elif isinstance(tool_output, Document):
                    all_external_docs.append(tool_output)
                elif isinstance(tool_output, list) and all(isinstance(doc, Document) for doc in tool_output):
                    all_external_docs.extend(tool_output)
                elif isinstance(tool_output, list) and all(isinstance(item, dict) and "page_content" in item for item in tool_output):
                    for item_dict in tool_output:
                        all_external_docs.append(Document(**item_dict))
                else:
                    print(f"Tool {tool_name} output not directly convertible to Document: {type(tool_output)}")

            except Exception as e:
                print(f"Error invoking external search tool {tool.name if hasattr(tool, 'name') else 'UnnamedTool'}: {e}")
        
        print(f"Found {len(all_external_docs)} documents from external search.")
        return {
            "documents": all_external_docs,
            "attempted_external_search": True,
            "document_assessment": None,
            "current_phase_iterations": state.get("current_phase_iterations", 0),
        }

    workflow = StateGraph(RagState)
    workflow.add_node("retrieve", retrieve_documents_node)
    if do_document_grading:
        workflow.add_node("grade_documents", grade_retrieved_documents_node)
    workflow.add_node("generate", generate_answer_node)
    workflow.add_node("transform_query", transform_query_node)
    if do_external_search and external_search_tools:
        workflow.add_node("external_search", perform_external_search_node)

    def route_after_retrieval(state: RagState) -> str:
        print(f"---ROUTING AFTER RETRIEVAL (iteration {state.get('iterations')}, phase iter {state.get('current_phase_iterations')})---")
        documents = state.get("documents")
        if documents:
            print("Documents found by primary retriever.")
            return "grade_documents" if do_document_grading else "generate"
        else:
            print("No documents found by primary retriever.")
            if do_external_search and external_search_tools and not state.get("attempted_external_search"):
                print("Attempting external search.")
                return "external_search"
            elif state.get("current_phase_iterations", 0) < MAX_QUERY_TRANSFORMATIONS:
                print("Attempting to transform query.")
                return "transform_query"
            else:
                print("Max transformations or no external search option. Ending.")
                return END

    def route_after_grading(state: RagState) -> str:
        print(f"---ROUTING AFTER GRADING (assessment: {state.get('document_assessment')})---")
        assessment = state.get("document_assessment")
        if assessment == "relevant":
            print("Documents relevant. Proceeding to generate.")
            return "generate"
        else:
            print(f"Documents assessment: {assessment}. Fallback options.")
            if do_external_search and external_search_tools and not state.get("attempted_external_search"):
                print("Attempting external search due to poor grading.")
                return "external_search"
            elif state.get("current_phase_iterations", 0) < MAX_QUERY_TRANSFORMATIONS:
                print("Attempting to transform query due to poor grading.")
                return "transform_query"
            else:
                print("Max transformations or no external search after poor grading. Ending.")
                return END

    def route_after_generation(state: RagState) -> str:
        print(f"---ROUTING AFTER GENERATION---")
        generated_answer = state.get("generation", "").lower()
        is_insufficient = (
            len(generated_answer) < 20 or
            any(phrase in generated_answer for phrase in ["don't know", "cannot answer", "not sure", "unable to find"])
        )
        if not is_insufficient:
            print("Answer seems sufficient. Ending.")
            return END
        else:
            print("Answer deemed insufficient.")
            if do_external_search and external_search_tools and not state.get("attempted_external_search"):
                print("Attempting external search due to insufficient answer.")
                return "external_search"
            elif state.get("current_phase_iterations", 0) < MAX_QUERY_TRANSFORMATIONS:
                print("Attempting to transform query due to insufficient answer.")
                return "transform_query"
            else:
                print("Max transformations or no external search after insufficient answer. Ending.")
                return END

    def route_after_transform_query(state: RagState) -> str:
        print(f"---ROUTING AFTER TRANSFORM QUERY (new query: '{state.get('question')}')---")
        return "retrieve"

    def route_after_external_search(state: RagState) -> str:
        print(f"---ROUTING AFTER EXTERNAL SEARCH---")
        documents_from_external_search = state.get("documents")
        if not documents_from_external_search:
            print("External search yielded no documents.")
            if state.get("current_phase_iterations", 0) < MAX_QUERY_TRANSFORMATIONS:
                print("Attempting to transform query as external search failed.")
                return "transform_query"
            else:
                print("Max transformations and external search also failed. Ending.")
                return END
        else:
            print("Documents found from external search.")
            return "grade_documents" if do_document_grading else "generate"

    workflow.set_entry_point("retrieve")
    workflow.add_conditional_edges("retrieve", route_after_retrieval,
        {
            "grade_documents": "grade_documents" if do_document_grading else "generate",
            "generate": "generate",
            "external_search": "external_search" if do_external_search and external_search_tools else "transform_query",
            "transform_query": "transform_query",
            END: END,
        }
    )

    if do_document_grading:
        workflow.add_conditional_edges("grade_documents", route_after_grading,
            {
                "generate": "generate",
                "external_search": "external_search" if do_external_search and external_search_tools else "transform_query",
                "transform_query": "transform_query",
                END: END,
            }
        )
    
    workflow.add_conditional_edges("generate", route_after_generation,
        {
            "external_search": "external_search" if do_external_search and external_search_tools else "transform_query",
            "transform_query": "transform_query",
            END: END,
        }
    )

    workflow.add_conditional_edges("transform_query", route_after_transform_query, {"retrieve": "retrieve"})

    if do_external_search and external_search_tools:
        workflow.add_conditional_edges("external_search", route_after_external_search,
            {
                "grade_documents": "grade_documents" if do_document_grading else "generate",
                "generate": "generate",
                "transform_query": "transform_query",
                END: END,
            }
        )

    return workflow.compile(checkpointer=checkpointer, debug=debug)
