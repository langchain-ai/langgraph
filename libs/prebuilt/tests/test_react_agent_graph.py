from typing import Callable, Union

import pytest
from pydantic import BaseModel
from syrupy import SnapshotAssertion

from langgraph.prebuilt import create_react_agent
from tests.model import FakeToolCallingModel

model = FakeToolCallingModel()


def tool() -> None:
    """Testing tool."""
    ...


def pre_model_hook() -> None:
    """Pre-model hook."""
    ...


def post_model_hook() -> None:
    """Post-model hook."""
    ...


class ResponseFormat(BaseModel):
    """Response format for the agent."""

    result: str


@pytest.mark.parametrize("tools", [[], [tool]])
@pytest.mark.parametrize("pre_model_hook", [None, pre_model_hook])
@pytest.mark.parametrize("post_model_hook", [None, post_model_hook])
@pytest.mark.parametrize("response_format", [None, ResponseFormat])
def test_react_agent_graph_structure(
    snapshot: SnapshotAssertion,
    tools: list[Callable],
    pre_model_hook: Union[Callable, None],
    post_model_hook: Union[Callable, None],
    response_format: Union[type[BaseModel], None],
) -> None:
    agent = create_react_agent(
        model,
        tools=tools,
        pre_model_hook=pre_model_hook,
        post_model_hook=post_model_hook,
        response_format=response_format,
    )
    assert agent.get_graph().draw_mermaid(with_styles=False) == snapshot
