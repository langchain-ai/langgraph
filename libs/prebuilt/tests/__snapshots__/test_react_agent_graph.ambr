# serializer version: 1
# name: test_react_agent_graph_structure[None-None-None-tools0]
  '''
  graph TD;
  	__start__ --> agent;
  	agent --> __end__;
  
  '''
# ---
# name: test_react_agent_graph_structure[None-None-None-tools1]
  '''
  graph TD;
  	__start__ --> agent;
  	agent -.-> __end__;
  	agent -.-> tools;
  	tools --> agent;
  
  '''
# ---
# name: test_react_agent_graph_structure[None-None-pre_model_hook-tools0]
  '''
  graph TD;
  	__start__ --> pre_model_hook;
  	pre_model_hook --> agent;
  	agent --> __end__;
  
  '''
# ---
# name: test_react_agent_graph_structure[None-None-pre_model_hook-tools1]
  '''
  graph TD;
  	__start__ --> pre_model_hook;
  	agent -.-> __end__;
  	agent -.-> tools;
  	pre_model_hook --> agent;
  	tools --> pre_model_hook;
  
  '''
# ---
# name: test_react_agent_graph_structure[None-post_model_hook-None-tools0]
  '''
  graph TD;
  	__start__ --> agent;
  	agent --> post_model_hook;
  	post_model_hook --> __end__;
  
  '''
# ---
# name: test_react_agent_graph_structure[None-post_model_hook-None-tools1]
  '''
  graph TD;
  	__start__ --> agent;
  	agent --> post_model_hook;
  	post_model_hook -.-> __end__;
  	post_model_hook -.-> agent;
  	post_model_hook -.-> tools;
  	tools --> agent;
  
  '''
# ---
# name: test_react_agent_graph_structure[None-post_model_hook-pre_model_hook-tools0]
  '''
  graph TD;
  	__start__ --> pre_model_hook;
  	agent --> post_model_hook;
  	pre_model_hook --> agent;
  	post_model_hook --> __end__;
  
  '''
# ---
# name: test_react_agent_graph_structure[None-post_model_hook-pre_model_hook-tools1]
  '''
  graph TD;
  	__start__ --> pre_model_hook;
  	agent --> post_model_hook;
  	post_model_hook -.-> __end__;
  	post_model_hook -.-> pre_model_hook;
  	post_model_hook -.-> tools;
  	pre_model_hook --> agent;
  	tools --> pre_model_hook;
  
  '''
# ---
# name: test_react_agent_graph_structure[ResponseFormat-None-None-tools0]
  '''
  graph TD;
  	__start__ --> agent;
  	agent --> generate_structured_response;
  	generate_structured_response --> __end__;
  
  '''
# ---
# name: test_react_agent_graph_structure[ResponseFormat-None-None-tools1]
  '''
  graph TD;
  	__start__ --> agent;
  	agent -.-> generate_structured_response;
  	agent -.-> tools;
  	tools --> agent;
  	generate_structured_response --> __end__;
  
  '''
# ---
# name: test_react_agent_graph_structure[ResponseFormat-None-pre_model_hook-tools0]
  '''
  graph TD;
  	__start__ --> pre_model_hook;
  	agent --> generate_structured_response;
  	pre_model_hook --> agent;
  	generate_structured_response --> __end__;
  
  '''
# ---
# name: test_react_agent_graph_structure[ResponseFormat-None-pre_model_hook-tools1]
  '''
  graph TD;
  	__start__ --> pre_model_hook;
  	agent -.-> generate_structured_response;
  	agent -.-> tools;
  	pre_model_hook --> agent;
  	tools --> pre_model_hook;
  	generate_structured_response --> __end__;
  
  '''
# ---
# name: test_react_agent_graph_structure[ResponseFormat-post_model_hook-None-tools0]
  '''
  graph TD;
  	__start__ --> agent;
  	agent --> post_model_hook;
  	post_model_hook --> generate_structured_response;
  	generate_structured_response --> __end__;
  
  '''
# ---
# name: test_react_agent_graph_structure[ResponseFormat-post_model_hook-None-tools1]
  '''
  graph TD;
  	__start__ --> agent;
  	agent --> post_model_hook;
  	post_model_hook -.-> agent;
  	post_model_hook -.-> generate_structured_response;
  	post_model_hook -.-> tools;
  	tools --> agent;
  	generate_structured_response --> __end__;
  
  '''
# ---
# name: test_react_agent_graph_structure[ResponseFormat-post_model_hook-pre_model_hook-tools0]
  '''
  graph TD;
  	__start__ --> pre_model_hook;
  	agent --> post_model_hook;
  	post_model_hook --> generate_structured_response;
  	pre_model_hook --> agent;
  	generate_structured_response --> __end__;
  
  '''
# ---
# name: test_react_agent_graph_structure[ResponseFormat-post_model_hook-pre_model_hook-tools1]
  '''
  graph TD;
  	__start__ --> pre_model_hook;
  	agent --> post_model_hook;
  	post_model_hook -.-> generate_structured_response;
  	post_model_hook -.-> pre_model_hook;
  	post_model_hook -.-> tools;
  	pre_model_hook --> agent;
  	tools --> pre_model_hook;
  	generate_structured_response --> __end__;
  
  '''
# ---
